### PR TITLE
[WM-2570] Implement task cost

### PR DESCRIPTION
**Pull request by kpierre13**
### What
- Adding the task cost onto the workflow details page for ToA

### Why
- Cost visibility

### Testing strategy

- [x] Unit tests

image
image

For failed tasks, we don't show any cost:
image

For running tasks, we tell the user's that the the task is currently in progress:
image

Existing workflow dashboard in GCP is unchanged:
image

[WM-2570]: https://broadworkbench.atlassian.net/browse/WM-2570?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/job-common.js [hunk 1]
```diff
@@ -161,3 +161,12 @@ export const workflowDetailsBreadcrumbSubtitle = (namespace, workspaceName, subm
     h3({ style: Style.elements.sectionHeader }, [`Workflow ${workflowId}`]),
   ]);
 };
+
+export const getTaskCost = (vmCost, startTime, endTime) => {
```
Comment (nawatts) [inline src/components/job-common.js]: An options object might be nice here, since there are several arguments to the function.
https://www.codereadability.com/what-are-javascript-options-objects/

File: src/pages/workspaces/workspace/jobHistory/CallTable.js [hunk 2]
```diff
@@ -346,6 +347,40 @@ const CallTable = ({
...
+                          },
+                          [icon('info-circle', { style: { marginLeft: '0.4rem', color: colors.accent(1) } })]
+                        ),
+                      ]),
+                    cellRenderer: ({ rowIndex }) => {
+                      const { vmCostUsd, taskStartTime, taskEndTime, executionStatus, callCaching } = filteredCallObjects[rowIndex];
+                      if (taskStartTime && vmCostUsd) {
+                        if (executionStatus === 'Running') {
+                          const cost = getTaskCost(vmCostUsd, taskStartTime);
+                          return div({}, [span({ style: { fontStyle: 'italic' } }, ['In Progress - ']), `$${cost}`]);
```
Comment (nawatts) [inline src/pages/workspaces/workspace/jobHistory/CallTable.js]: Nit: empty props objects can be omitted.

```suggestion
                          return div([span({ style: { fontStyle: 'italic' } }, ['In Progress - ']), `$${cost}`]);
```

File: src/workflows-app/RunDetails.js [hunk 3]
```diff
@@ -151,6 +157,33 @@ export const BaseRunDetails = (
...
+    });
+    return total;
+  };
+
+  useEffect(() => {
+    if (callObjects) {
+      const cost = calculateTotalCost(callObjects);
+      setTaskCostTotal(cost);
+    }
+  }, [callObjects]);
```
Comment (nawatts) [inline src/workflows-app/RunDetails.js]: It looks like `useMemo` might be preferable over `useEffect` here.

https://react.dev/reference/react/useMemo

```suggestion
  const taskCostTotal = useMemo(() => {
    return callObjects ? calculateTotalCost(callObjects) : undefined;
  }, [callObjects]);
```

File: src/workflows-app/RunDetails.js [hunk 4]
```diff
@@ -151,6 +157,33 @@ export const BaseRunDetails = (
     [signal, workspaceId]
   );
 
+  const inProgressElement = (workflow) => {
```
Comment (nawatts) [inline src/workflows-app/RunDetails.js]: Nit: for convention, suggest naming functions that return React elements `render...`.

```suggestion
  const renderInProgressElement = (workflow) => {
```

File: src/workflows-app/RunDetails.js [hunk 5]
```diff
@@ -151,6 +157,30 @@ export const BaseRunDetails = (
...
+  const renderInProgressElement = (workflow) => {
+    if (workflow.status === 'Running') {
+      return span({ style: { fontStyle: 'italic' } }, ['In progress - ']);
+    }
+  };
+
+  const calculateTotalCost = (callObjects) => {
+    let total = 0;
+    Object.values(callObjects).forEach((call) => {
+      if (!call[0].taskStartTime) {
```
Comment (cjllanwarne) [inline src/workflows-app/RunDetails.js]: Why is there an index inside a call? My guess is that this Is this because of shards of a scatter and automated retries

So... I assume we want to sum all of the shards and retries, not just take the first one?

File: src/workflows-app/RunDetails.test.js [hunk 6]
```diff
@@ -732,4 +755,4 @@ describe('BaseRunDetails - render smoke test', () => {
     // Ensure the diff is rendered
     screen.getByText('Result: View cache diff');
   });
-});
+}, 8000);
```
Comment (cjllanwarne) [inline src/workflows-app/RunDetails.test.js]: What's 8000?

File: src/workflows-app/RunDetails.js (hunk 5 above)
Comment (kpierre13) [reply src/workflows-app/RunDetails.js]: Yes, you're correct. This is updated to be more accurate in the work for my next ticket: https://broadworkbench.atlassian.net/browse/WM-2648

File: src/workflows-app/RunDetails.test.js (hunk 6 above)
Comment (kpierre13) [reply src/workflows-app/RunDetails.test.js]: This is a timer for the test. I actually don't recall adding this, but I'll remove and see if it was actually necessary.

File: src/components/job-common.js [hunk 7]
```diff
@@ -161,3 +161,12 @@ export const workflowDetailsBreadcrumbSubtitle = (namespace, workspaceName, subm
...
+  const currentEndTime = Date.parse(endTime) || Date.now();
+  const vmCostDouble = parseFloat(vmCost);
+  const startDateTime = Date.parse(startTime);
+
+  const elapsedTime = currentEndTime - startDateTime;
+  return parseFloat(((elapsedTime / 3600000) * vmCostDouble).toFixed(2));
```
Comment (jgainerdewar) [inline src/components/job-common.js:171]: Should we round this up to one cent when the cost is great than 0?

File: src/pages/workspaces/workspace/jobHistory/CallTable.js [hunk 8]
```diff
@@ -346,6 +347,37 @@ const CallTable = ({
...
+                    size: { basis: 200, grow: 1 },
+                    field: 'cost',
+                    headerRenderer: () =>
+                      h(Sortable, { sort, field: 'cost', onSort: setSort }, [
+                        'Approximate Cost',
+                        h(
+                          TooltipTrigger,
+                          {
+                            content:
+                              'This is a calculated list price and does not include disk cost, subworkflow costs, or any cloud account discounts.',
```
Comment (jgainerdewar) [inline src/pages/workspaces/workspace/jobHistory/CallTable.js]: Might be clearer to say: "Approximate cost is calculated based on the the list price of the VM used and does not include disk cost or any cloud account discounts."

Reasons:
 * In this table we see the subworkflows and their missing costs, there is no total number they're excluded from
 * The number shown isn't a published list price, the list price is one input to it.

File: src/workflows-app/RunDetails.js [hunk 9]
```diff
@@ -249,10 +278,22 @@ export const BaseRunDetails = (
...
             h(TroubleshootingBox, { name, namespace, logUri: workflow.workflowLog, submissionId, workflowId, showLogModal }, []),
           ]),
+          div({ style: { fontSize: 16, padding: '0rem 2.5rem 1rem' } }, [
+            span({ style: { fontWeight: 'bold' } }, ['Approximate workflow cost: ']),
+            renderInProgressElement(workflow),
+            `$${taskCostTotal.toFixed(2)}`,
+            h(
+              TooltipTrigger,
+              {
+                content: 'This total is an estimate rounded up to the nearest cent and does not include subworkflow costs.',
```
Comment (jgainerdewar) [inline src/workflows-app/RunDetails.js]: Is it rounded up to the nearest cent? It doesn't look like that's what `toFixed` does. I think it would be good to make this consistent with the other tooltip - if we do want to round everything UP to the next cent, say that in both places. 

Suggested edit (assuming no change to rounding): "Approximate cost is calculated based on the the list price of the VMs used and does not include disk cost, subworkflow cost, or any cloud account discounts."

In this one we DO want to mention subworkflow cost, since it's a sum that users might expect to include subworkflow tasks.

File: src/pages/workspaces/workspace/jobHistory/CallTable.js [hunk 10]
```diff
@@ -346,6 +347,37 @@ const CallTable = ({
...
+                      const { vmCostUsd, taskStartTime, taskEndTime, executionStatus, callCaching, subWorkflowId } = filteredCallObjects[rowIndex];
+                      if (taskStartTime && vmCostUsd) {
+                        if (executionStatus === 'Running') {
+                          const cost = getTaskCost(vmCostUsd, taskStartTime);
+                          return div([span({ style: { fontStyle: 'italic' } }, ['In Progress - ']), `$${cost}`]);
+                        }
+                        const cost = getTaskCost(vmCostUsd, taskStartTime, taskEndTime);
+                        return div({}, [`$${cost}`]);
+                      }
+                      if (executionStatus === 'Failed' || callCaching?.hit === true || !_.isEmpty(subWorkflowId)) {
```
Comment (jgainerdewar) [inline src/pages/workspaces/workspace/jobHistory/CallTable.js]: Why don't we want to show cost for failed tasks?

File: src/components/job-common.js (hunk 7 above)
Comment (jgainerdewar) [reply src/components/job-common.js:171]: I think the best behavior here would be to round (up or down) to the nearest cent, and if that turns out to be 0.00 instead display `< $0.01`

File: src/pages/workspaces/workspace/jobHistory/CallTable.js (hunk 10 above)
Comment (kpierre13) [reply src/pages/workspaces/workspace/jobHistory/CallTable.js]: Here we should still be showing the cost for failed tasks. This is in the event that a task fails before we can get a start/vm cost from TES. Without it, tasks that are failed with no additional data remain in a `Fetching cost information` state
image

File: src/pages/workspaces/workspace/jobHistory/CallTable.js [hunk 11]
```diff
@@ -346,6 +347,40 @@ const CallTable = ({
...
+                    size: { basis: 200, grow: 1 },
+                    field: 'cost',
+                    headerRenderer: () =>
+                      h(Sortable, { sort, field: 'cost', onSort: setSort }, [
+                        'Approximate Cost',
+                        h(
+                          TooltipTrigger,
+                          {
+                            content:
+                              'Approximate cost is a calculated based on the list price of the VM, and does not include disk cost or any cloud account discounts.',
```
Comment (jgainerdewar) [inline src/pages/workspaces/workspace/jobHistory/CallTable.js]: You've got an extra "a" in there :)

File: src/workflows-app/RunDetails.js [hunk 12]
```diff
@@ -151,6 +156,37 @@ export const BaseRunDetails = (
     [signal, workspaceId]
   );
 
+  const renderInProgressElement = (workflow) => {
+    if (workflow.status === 'Running') {
+      return span({ style: { fontStyle: 'italic' } }, ['In progress - ']);
+    }
+  };
+
+  const renderTaskCostElement = (cost) => {
```
Comment (jgainerdewar) [inline src/workflows-app/RunDetails.js]: Can we apply some DRY and move this function to a place the call table can also use it?

File: src/workflows-app/RunDetails.js [hunk 13]
```diff
@@ -249,10 +285,23 @@ export const BaseRunDetails = (
...
           ]),
+          div({ style: { fontSize: 16, padding: '0rem 2.5rem 1rem' } }, [
+            span({ style: { fontWeight: 'bold' } }, ['Approximate workflow cost: ']),
+            renderInProgressElement(workflow),
+            renderTaskCostElement(taskCostTotal),
+            h(
+              TooltipTrigger,
+              {
+                content:
+                  'Approximate cost is calculated based on the the list price of the VMs used and does not include disk cost, subworkflow cost, or any cloud account discounts',
```
Comment (jgainerdewar) [inline src/workflows-app/RunDetails.js]: Double "the" here

**Review by jgainerdewar — APPROVED** [review]

File: src/components/job-common.js (hunk 1 above)
Comment (mspector) [reply src/components/job-common.js]: > An options object might be nice here, since there are several arguments to the function.
https://www.codereadability.com/what-are-javascript-options-objects/

+1 to Nick's suggestion

File: src/pages/workspaces/workspace/jobHistory/CallTable.js [hunk 14]
```diff
@@ -346,6 +347,37 @@ const CallTable = ({
...
+                      ]),
+                    cellRenderer: ({ rowIndex }) => {
+                      const { vmCostUsd, taskStartTime, taskEndTime, executionStatus, callCaching, subWorkflowId } = filteredCallObjects[rowIndex];
+                      if (taskStartTime && vmCostUsd) {
+                        if (executionStatus === 'Running') {
+                          const cost = getTaskCost(vmCostUsd, taskStartTime);
+                          return div([span({ style: { fontStyle: 'italic' } }, ['In Progress - ']), `$${cost}`]);
+                        }
+                        const cost = getTaskCost(vmCostUsd, taskStartTime, taskEndTime);
+                        return div({}, [renderTaskCostElement(cost)]);
```
Comment (cjllanwarne) [inline src/pages/workspaces/workspace/jobHistory/CallTable.js]: ~I know it's "equivalent" but to make this easier (for me!) to parse, I might suggest putting this into an `else`~

File: src/pages/workspaces/workspace/jobHistory/CallTable.js [hunk 15]
```diff
@@ -346,6 +347,37 @@ const CallTable = ({
...
+                            content:
+                              'Approximate cost is calculated based on the list price of the VM, and does not include disk cost or any cloud account discounts.',
+                          },
+                          [icon('info-circle', { style: { marginLeft: '0.4rem', color: colors.accent(1) } })]
+                        ),
+                      ]),
+                    cellRenderer: ({ rowIndex }) => {
+                      const { vmCostUsd, taskStartTime, taskEndTime, executionStatus, callCaching, subWorkflowId } = filteredCallObjects[rowIndex];
+                      if (taskStartTime && vmCostUsd) {
+                        if (executionStatus === 'Running') {
```
Comment (cjllanwarne) [inline src/pages/workspaces/workspace/jobHistory/CallTable.js]: There's a lot of if-nesting here, and I have questions like "but what about state X" (eg "aborting"). I also had a similar question to Janet below. I wonder, would it be easier to use `if (endTime)` here instead of `if (executionStatus == 'Running')`? And then have something like - 

```js
// taskHasACost could be a testable function, equivalent to 'tastStartTime && vmCostUsd'
if (taskHasACost) {
  if (taskEndTime) {
    // Calculate with end time
  } else { 
    // Calculate without end time
  }
} else {
  // noCostExpected could be a separate function with its own unit tests and clear semantics 
  if (noCostExpected) {
    return '$0.00'
  } else {
    return 'Cost pending'
  }
}
```

File: src/workflows-app/RunDetails.js (hunk 5 above)
Comment (cjllanwarne) [reply src/workflows-app/RunDetails.js]: I don't think this is a subworkflows thing, this is a "plain old scattered task" thing. I worry that not only are we "not including subworkflows", we are also "ignoring every task in scatters except the first index".

Luckily, I think you can easily fix this by just iterating over all indexes in the `call` array, like eg
```js
 Object.values(callObjects).forEach((call) => {
    call.forEach((shard) => {
      if (!shard.taskStartTime) {
      [...]
```

File: src/workflows-app/RunDetails.js [hunk 16]
```diff
@@ -151,6 +156,30 @@ export const BaseRunDetails = (
...
+      } else if (call[0].taskEndTime) {
+        total += getTaskCost(call[0].vmCostUsd, call[0].taskStartTime, call[0].taskEndTime);
+      } else {
+        total += getTaskCost(call[0].vmCostUsd, call[0].taskStartTime);
+      }
+    });
+    return total;
+  };
+
+  const taskCostTotal = useMemo(() => {
```
Comment (cjllanwarne) [inline src/workflows-app/RunDetails.js:174]: Is the `useMemo` doing anything here? It looks like it would only be called once per page-load anyway?

File: src/workflows-app/RunDetails.js [hunk 17]
```diff
@@ -249,10 +278,23 @@ export const BaseRunDetails = (
...
+            span({ style: { fontWeight: 'bold' } }, ['Approximate workflow cost: ']),
+            renderInProgressElement(workflow),
+            renderTaskCostElement(taskCostTotal),
+            h(
+              TooltipTrigger,
+              {
+                content:
+                  'Approximate cost is calculated based on the list price of the VMs used and does not include disk cost, subworkflow cost, or any cloud account discounts',
+              },
+              [icon('info-circle', { style: { marginLeft: '0.4rem', color: colors.accent(1) } })]
```
Comment (cjllanwarne) [inline src/workflows-app/RunDetails.js:311]: Given the lack of subworkflows, I might advocate for making this more "warning" than "info-circle". What do you think?

File: src/workflows-app/RunDetails.test.js [hunk 18]
```diff
@@ -242,6 +242,29 @@ describe('BaseRunDetails - render smoke test', () => {
     });
   });
 
+  it('shows expected cost details', async () => {
```
Comment (cjllanwarne) [inline src/workflows-app/RunDetails.test.js:246]: These rendering tests are fine, but there are many cases that the `getTaskCost` and `calculateTotalCost` functions handle that could be unit-tested. That would also give us an easier framework to validate edge cases without having to wrap everything into a full page-render test.

File: src/pages/workspaces/workspace/jobHistory/CallTable.js (hunk 14 above)
Comment (cjllanwarne) [reply src/pages/workspaces/workspace/jobHistory/CallTable.js]: Actually, superseded by my below comment...

**Review by mspector — APPROVED** [review]

File: src/workflows-app/RunDetails.js (hunk 16 above)
Comment (kpierre13) [reply src/workflows-app/RunDetails.js:174]: Yeah, so from my understanding this is a way of caching the result on each re-render. It'll only run the function if its dependency, `callObjects` changes. That way we're not running this function on every re-render/page-refresh, only when necessary.

Comment (kpierre13) [reply src/workflows-app/RunDetails.js:174]: Also in response to shards, updating the code per your suggestion didn't seem to work? It looks like `shards` are nested as sub-workflows, so when they appear at the top level of the page, they give an option to `View Subworkflow` until you navigate down and eventually get the cost of that task (next slice will focus on bubbling this cost up each level) 
image
image

File: src/workflows-app/RunDetails.js (hunk 17 above)
Comment (kpierre13) [reply src/workflows-app/RunDetails.js:311]: I'm torn. My hesitation is around not wanting users to feel like they've done something wrong, or that something is wrong with their tasks/subworkflows. I think for that reason I'd go for the info-circle, but I can be convinced.
